packages/opentelemetry-scope-base/src/context.ts

```diff
@@ -0,0 +1,72 @@
+/*!
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Map of identifiers to an unknown value used internally to store context */
+type Store = { [identifer: string]: unknown };
+
+/**
+ * Class which stores and manages current context values. All methods which
+ * update context such as get and delete do not modify an existing context,
+ * but create a new one with updated values.
+ */
+export class Context {
+  private _context: Store;
+
+  public static readonly ROOT_CONTEXT = new Context();
+
+  /**
+   * Construct a new context which inherits values from an optional parent context.
+   *
+   * @param context a context from which to inherit values
+   */
+  protected constructor(parentContext: Store = {}) {
```

Comment:
private instead protected? e.g. `setValue` will call this constructor so even if someone extends this class the result of `setValue` will be the base type.

Reply:
You can't extend a class with a private constructor because you can't call `super`

Reply:
Noone should extend it. Or what is the use case of extending it?
I think if user defined Context classes should be supported an abstract API like `createContext()` would be needed.

Reply:
I extended it in `core` to add convenience methods. I guess they could just be made to be functions that take `Context`

Reply:
> I extended it in `core` to add convenience methods. I guess they could just be made to be functions that take `Context`

updated